### PR TITLE
Add `--tree` option to `dds ls`

### LINKED ISSUE #111: List whole directory structure
At the moment the cli `dds ls [projectID]` command lists the root level folders and files. Add possibility of listing entire directory structure and use the pagination functionality.

### PR DESCRIPTION
**Pull request by ErikDanielsson**
Added `--tree` option to `dds ls` for showing the whole file tree in the remote. Should close #111. Showing file sizes depends on merging https://github.com/ScilifelabDataCentre/dds_web/pull/547.

### Note about implementation
The method uses two tree walks, one for constructing the file tree and computing the maximum indentation needed, and one for the construction of the `rich` tree. These could however be coalesced into one by updating the subtrees recursively if one finds that the maximum indentation needed for one subtree is greater than the other subtrees. I didn't implement it this way for readability (and ease of implementation) reasons. If it happens that this command is slow with the current implementation however, the improved speed of using one tree walk could perhaps justify the added complexity.

### Demo
```console
$ dds ls fac002 --tree --size
     ︵ 
 ︵ (  )   ︵ 
(  ) ) (  (  )   SciLifeLab Data Delivery System 
 ︶  (  ) ) (    https://www.scilifelab.se/data 
      ︶ (  )    Version 0.1.dev0 
          ︶

                                                                                                                                                                                                                   
 Files & directories in project: fac002/                                                                                                                                                                           
 ├── .bashrc                                             5.025      KB                                                                                                                                             
 └── dds_subfolders/                                                                                                                                                                                               
     ├── f1/                                                                                                                                                                                                       
     │   ├── d3/                                                                                                                                                                                                   
     │   │   └── d4/                                                                                                                                                                                               
     │   │       └── d5/                                                                                                                                                                                           
     │   │           └── d6/                                                                                                                                                                                       
     │   │               └── d7/                                                                                                                                                                                   
     │   │                   └── d8/                                                                                                                                                                               
     │   │                       └── d9/                                                                                                                                                                           
     │   │                           └── d10/                                                                                                                                                                      
     │   │                               └── file.txt    4.000   bytes                                                                                                                                             
     │   └── file.txt                                    2.000   bytes                                                                                                                                             
     ├── f2/                                                                                                                                                                                                       
     │   └── file.txt                                    2.000   bytes                                                                                                                                             
     └── f3/                                                                                                                                                                                                       
         └── file.txt                                    2.000   bytes  
```

### CODE REVIEW

File: dds_cli/data_lister.py [hunk 1]
```diff
@@ -307,3 +319,130 @@ def list_files(self, folder: str = None, show_size: bool = False):
...
+            # Get max length of size string
+            max_size = max(
+                [
+                    len(x["size"].split(" ")[0])
+                    for x in sorted_files_folders
+                    if show_size and "size" in x
+                ],
+                default=0,
+            )
+            # Rich outputs precisely one line per
```
Comment (ErikDanielsson) [inline dds_cli/data_lister.py:367]:
```suggestion
            # Rich outputs precisely one line per file/folder
```

File: dds_cli/data_lister.py [hunk 2]
```diff
@@ -307,3 +319,130 @@ def list_files(self, folder: str = None, show_size: bool = False):
...
         return visible_folders
+
+    def list_recursive(self, show_size: bool = False):
+        @dataclass
+        class FileTree:
+            """
+            Container class for holding information about the remote file tree
+            """
+
+            subtrees: list[Union["FileTree", Tuple[str, str]]] = None
```
Comment (inaod568) [inline dds_cli/data_lister.py:330]: I get the following traceback
```
  File "/Users/inaod568/repos/DDS/dds_cli/dds_cli/__main__.py", line 312, in ls
    lister.list_recursive(show_size=size)
  File "/Users/inaod568/repos/DDS/dds_cli/dds_cli/data_lister.py", line 325, in list_recursive
    class FileTree:
  File "/Users/inaod568/repos/DDS/dds_cli/dds_cli/data_lister.py", line 330, in FileTree
    subtrees: list[Union["FileTree", Tuple[str, str]]] = None
TypeError: 'type' object is not subscriptable
```

**Review by inaod568 — CHANGES_REQUESTED** [review]

File: dds_cli/data_lister.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ErikDanielsson) [reply dds_cli/data_lister.py]: Didn't know this, thanks! Weird that it executed correctly on my machine.

**Review by inaod568 — APPROVED** [review]
Looks good, I like it!

File: dds_cli/data_lister.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (inaod568) [reply dds_cli/data_lister.py]: Yeah, I'm not sure why it's behaving differently!

**Pull request merged**